cpp/src/stream_compaction/distinct_count.cu

```diff
@@ -76,97 +55,235 @@ struct check_for_nan {
+    // TODO: cuco::make_pair currently requires rvalue references. We
+    // create a copy to avoid double-move invoking undefined behavior.
+    auto ii = i;
+    return cuco::make_pair(std::move(i), std::move(ii));
+  });
+
+  // when nulls are equal, insert non-null rows only to improve efficiency
+  if (nulls_equal == null_equality::EQUAL and has_null) {
+    thrust::counting_iterator<size_type> stencil(0);
+    auto const [row_bitmask, null_count] = cudf::detail::bitmask_or(keys, stream);
+    row_validity pred{static_cast<bitmask_type const*>(row_bitmask.data())};
+
+    key_map.insert_if(iter, iter + num_rows, stencil, pred, hash_key, row_equal, stream.value());
+    return key_map.get_size() + static_cast<std::size_t>((null_count > 0) ? 1 : 0);
+  }
+  // otherwise, insert all
+  key_map.insert(iter, iter + num_rows, hash_key, row_equal, stream.value());
+  return key_map.get_size();
+}
+
 cudf::size_type distinct_count(column_view const& input,
                                null_policy null_handling,
                                nan_policy nan_handling,
                                rmm::cuda_stream_view stream)
 {
-  if (0 == input.size() || input.null_count() == input.size()) { return 0; }
+  auto const num_rows = input.size();
+
+  if (num_rows == 0 or num_rows == input.null_count()) { return 0; }
 
-  cudf::size_type nrows = input.size();
+  auto const count_nulls      = null_handling == null_policy::INCLUDE;
+  auto const nan_is_null      = nan_handling == nan_policy::NAN_IS_NULL;
+  auto const should_check_nan = cudf::type_dispatcher(input.type(), check_nan_predicate{});
+  auto input_device_view      = cudf::column_device_view::create(input, stream);
+  auto device_view            = *input_device_view;
+  auto input_table_view       = table_view{{input}};
+  auto table_ptr              = cudf::table_device_view::create(input_table_view, stream);
+  row_equality_comparator comp(nullate::DYNAMIC{cudf::has_nulls(input_table_view)},
+                               *table_ptr,
+                               *table_ptr,
+                               null_equality::EQUAL);
+
+  return thrust::count_if(
+    rmm::exec_policy(stream),
+    thrust::counting_iterator<cudf::size_type>(0),
+    thrust::counting_iterator<cudf::size_type>(num_rows),
+    [count_nulls, nan_is_null, should_check_nan, device_view, comp] __device__(cudf::size_type i) {
+      bool is_nan  = false;
+      bool is_null = device_view.is_null(i);
+      if (nan_is_null and should_check_nan) {
+        is_nan = cudf::type_dispatcher(device_view.type(), check_nan{}, device_view, i);
+      }
+      if (count_nulls) {
+        if (nan_is_null and (is_nan or is_null)) {
+          auto prev_is_nan =
+            should_check_nan
+              ? cudf::type_dispatcher(device_view.type(), check_nan{}, device_view, i - 1)
+              : false;
+          return (i == 0 or not(device_view.is_null(i - 1) or prev_is_nan));
+        }
```

Comment:
Fusing into one if statement cannot work here cause we have an `else if` statement right after depending on `count_nulls`.

Reply:
Oh, I see. You can fuse them and add `not count_nulls and ...` to the `else if`. The current nesting is much harder to read than it would be if the conditions were combined. It would be like this:

```diff
-      if (count_nulls) {
-        if (nan_is_null and (is_nan or is_null)) {
-          auto prev_is_nan =
-            should_check_nan
-              ? cudf::type_dispatcher(device_view.type(), check_nan{}, device_view, i - 1)
-              : false;
-          return (i == 0 or not(device_view.is_null(i - 1) or prev_is_nan));
-        }
-      } else {
-        if (is_null or (nan_is_null and is_nan)) { return false; }
-      }
+      if (count_nulls and nan_is_null and (is_nan or is_null)) {
+       auto prev_is_nan =
+          should_check_nan
+            ? cudf::type_dispatcher(device_view.type(), check_nan{}, device_view, i - 1)
+            : false;
+        return (i == 0 or not(device_view.is_null(i - 1) or prev_is_nan));
+      } else if (not count_nulls and (is_null or (nan_is_null and is_nan))) {
+        return false;
+      }
```
However, I think it would be worthwhile to work through this boolean algebra a little more and see if we can reduce its complexity. In particular, the `i==0` case could be handled as an early-exit condition. We should not allow the computation of `prev_is_nan` if `i==0` because it fetches data out of bounds.